### PR TITLE
Draft aria-* and data-* additions for Bootstrap

### PR DESCRIPTION
**Pull request by rc1405**
Pull request to discuss implementation of aria-* and data-* options for bootstrap and other uses.  Proposed solution to #78

### CODE REVIEW

**Comment by NMRobert** [comment]
This is working for me - it might be good to alter it so that it affects dcc.* components as well - currently they don't throw an exception, but the attributes don't show up.

File: dash/development/base_component.py [hunk 1]
```diff
@@ -223,7 +226,7 @@ def __init__(self, {default_argtext}):
 
         def __repr__(self):
             if(any(getattr(self, c, None) is not None for c in self._prop_names
-                   if c is not self._prop_names[0])):
+                   if (c is not self._prop_names[0] and (not c[:5] in ['data_', 'aria_'])))):
```
Comment (chriddyp) [inline dash/development/base_component.py]: `__repr__` should print out a string that can be copied and pasted back into a python program to get the same result. In this case, it looks like we have:
```
>>> html.Div('test', data_foo='bar')
Div(children='test', data-foo='bar')
```

**Comment by chriddyp** [comment]
>  it might be good to alter it so that it affects dcc.* components as well 

That's a good point. The `dash_html_components` will end up rendering these attributes but any non-HTML component will not. In those cases, those components should throw an exception if a property wasn't supplied and if the component doesn't accept it.

I'm not sure what the solution would be for that case. Perhaps the component authors can include a special prop-type like `'data-*': PropTypes.string` which would have a special interpretation by the dash abstract component.

File: dash/development/base_component.py (hunk 1 above)
Comment (rc1405) [inline dash/development/base_component.py]: __repr__ adjusted to provide output as described

Comment (rc1405) [inline dash/development/base_component.py]: Output is correct at
`Div(children='test', data_foo='bar')`

**Comment by chriddyp** [comment]
> I'm not sure what the solution would be for that case. Perhaps the component authors can include a special prop-type like 'data-*': PropTypes.string which would have a special interpretation by the dash abstract component.

As mentioned above, I think that we'll need a solution that allows general component authors to specify that these properties are valid, rather than monkey patching the base component.
I wrote a guide in the community forum on the steps required to make this happen: https://community.plot.ly/t/can-data-attributes-be-created-in-dash/7222/2?u=chriddyp

I'm going to close this for now in favor of a more comprehensive approach. Thanks for opening the discussion!

**Pull request closed without merging**